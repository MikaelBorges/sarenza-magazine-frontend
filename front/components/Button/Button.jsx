import React from 'react';
import styled from 'styled-components';

const Link = styled.a`
    color: #fff;
    z-index: 0;
    position: relative;
    vertical-align: middle;
    font: 700 12px/14px robotoc, Arial Narrow, sans-serif;
    text-transform: uppercase;
    padding: 0 64px 0 32px;
    min-width: 96px;
    height: 48px;
    overflow: hidden;
    outline: 0;
    transition: background 0.25s ease, color 0.25s ease, border 0.25s ease;
    display: inline-flex;
    align-items: center;
    text-align: left;
    text-decoration: none;
    margin: 0 auto 40px auto;
    background: #1d1d1b;

    &:visited,
    &:hover {
        text-decoration: none;
        color: #fff;
    }

    &:after {
        position: absolute;
        right: 32px;
        top: 14px;
        margin-top: 0;
        margin-left: 1.5em;
        content: '\\e372';
        font-family: pictenza;
        font-size: 1em;
        font-style: normal;
        speak: none;
        font-weight: 700;
        -webkit-font-smoothing: antialiased;
    }
`;
export const Button = ({ action, value, className }) => {
    return value ? (
        <Link href={action} className={className}>
            {value}
        </Link>
    ) : (
        'Erreur'
    );
};
