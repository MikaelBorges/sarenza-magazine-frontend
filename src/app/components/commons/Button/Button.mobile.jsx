import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';

import Icon from '../Icon';
import * as ICONS from '../Icon/constants';
import Loading, { ANIM_TYPE } from '../Loader/Loading';
import Status from '../Status';
import STATUS_TYPE from '../Status/constants';
import styles from './Button.mobile.module.scss';

const Button = ({
  disabled,
  onClick,
  onKeyDown,
  onFocus,
  onBlur,
  onEnter,
  onLeave,
  customClass,
  iconName,
  iconWidth,
  iconHeight,
  iconAfter,
  title,
  children,
  dataTestid,
  isLoading,
  isLoaded,
  stopLoaderFn,
  whiteTheme
}) => {
  const childrenLabel = children && <span className={styles.buttonLabel}>{children}</span>;
  // eslint-disable-next-line react/prop-types
  const ButtonIcon = ({ cssClass }) => (
    <Icon name={iconName} width={iconWidth} height={iconHeight} cssClass={cssClass} />
  );
  const chidrenWithIcon = iconName ? (
    <>
      {iconAfter ? (
        <>
          {childrenLabel}
          <ButtonIcon cssClass={styles.iconAfter} />
        </>
      ) : (
        <>
          <ButtonIcon cssClass={styles.iconBefore} />
          {childrenLabel}
        </>
      )}
    </>
  ) : (
    <>{childrenLabel}</>
  );

  const cssClasses = classnames(styles.button, styles.default, customClass, {
    [styles.disabled]: disabled,
    [styles.hasLoader]: isLoading || isLoaded,
    [styles.whiteTheme]: whiteTheme
  });

  useEffect(() => {
    if (isLoaded) {
      setTimeout(() => stopLoaderFn(), 3000);
    }
  }, [isLoaded, stopLoaderFn]);

  if (onClick || onKeyDown) {
    // eslint-disable-next-line no-nested-ternary
    const buttonStatus = isLoading
      ? STATUS_TYPE.WARNING
      : isLoaded
      ? STATUS_TYPE.SUCCESS
      : undefined;
    return (
      <Status
        type={buttonStatus}
        renderElement={(statusCssClasses) => (
          <button
            type="button"
            title={title}
            onFocus={onFocus}
            onMouseEnter={onEnter}
            onMouseLeave={onLeave}
            onBlur={onBlur}
            onKeyDown={onKeyDown}
            onClick={!isLoading && !isLoaded ? onClick : () => {}}
            disabled={disabled}
            className={cssClasses.concat(' ', statusCssClasses)}
            data-testid={dataTestid}>
            {isLoading && (
              <>
                <Loading
                  noPadding
                  animationType={ANIM_TYPE.SPINNER}
                  cssClass={styles.iconBefore}
                  size="16px"
                />
                <span className={styles.buttonLabel}>Chargement...</span>
              </>
            )}
            {isLoaded && (
              <>
                <Icon name={ICONS.CHECKED} cssClass={styles.iconBefore} width="16px" />
                <span className={styles.buttonLabel}>C&rsquo;est bon !</span>
              </>
            )}
            {!isLoading && !isLoaded && chidrenWithIcon}
          </button>
        )}
      />
    );
  }

  return <>Lien en erreur</>;
};

Button.propTypes = {
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onKeyDown: PropTypes.func,
  onBlur: PropTypes.func,
  onEnter: PropTypes.func,
  onLeave: PropTypes.func,
  customClass: PropTypes.string,
  iconName: PropTypes.string,
  iconWidth: PropTypes.string,
  iconHeight: PropTypes.string,
  iconAfter: PropTypes.bool,
  children: PropTypes.node,
  title: PropTypes.string,
  dataTestid: PropTypes.string,
  isLoading: PropTypes.bool,
  isLoaded: PropTypes.bool,
  stopLoaderFn: PropTypes.func
};

Button.defaultProps = {
  disabled: false,
  onClick: undefined,
  onFocus: undefined,
  onKeyDown: undefined,
  onBlur: undefined,
  onEnter: undefined,
  onLeave: undefined,
  customClass: null,
  iconName: undefined,
  iconWidth: '10px',
  iconHeight: '12px',
  iconAfter: false,
  children: undefined,
  title: '',
  dataTestid: '',
  isLoading: false,
  isLoaded: false,
  stopLoaderFn: () => {}
};

export default Button;
