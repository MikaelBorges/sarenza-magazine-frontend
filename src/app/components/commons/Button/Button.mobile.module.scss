@import '../../_styles/theme';

$button-color: $color-primary !default;
$button-bg: #fff !default;
$button-width: 50px !default;
$button-padding: 6px 8px !default;
$button-border-color: $color-border !default;
$button-border-width: 1px !default;
$button-border-radius: 4px !default;
$button-hover-bg: $color-primary !default;
$button-hover-color: $color-white !default;

.button {
  box-sizing: border-box;
  white-space: nowrap;
  display: inline-block;
  position: relative;
  margin: 0;
  line-height: 1.75;
  text-decoration: none;
  border: none;
  outline: none;
  cursor: default;

  &:not(:disabled) {
    &:hover {
      cursor: pointer;
    }
  }

  &.hasLoader:not(:disabled):hover {
    cursor: default;
  }

  .iconAfter {
    margin-left: 5px;
    position: absolute;
    height: top;
    top: 50%;
    transform: translate(-90%, -50%);
    left: 90%;
  }
  .iconBefore {
    margin-right: 5px;
  }

  &,
  svg {
    //transition: all 0.2s ease-in;
  }
}

.default {
  width: 192px;
  height: 48px;
  font-size: 12px;
  padding: 16px 32px;
  text-transform: uppercase;
  background-color: #000000;
  color: #fff;
  font-weight: bold;
  font-family: RobotoC;

  &.whiteTheme {
    background-color: white;
    color: #1d1d1b;
  }

  span {
    font-size: 12px;
  }
  svg {
    fill: $button-color;
  }

  &:not(:disabled) {
    &:hover,
    &:active {
      // background: $button-hover-bg;
      // color: $button-hover-color;
      cursor: pointer;

      svg {
        fill: $button-hover-color;
      }
    }
    &:active {
      // box-shadow: inset 1px 1px 1px rgba(0, 0, 0, 0.3);
    }
  }
}

.cta {
  @extend .default;
  background-color: $color-primary;
  color: #fff;
}

.disabled {
  opacity: 0.4;
  color: $color-text-light;
  cursor: not-allowed;
}
