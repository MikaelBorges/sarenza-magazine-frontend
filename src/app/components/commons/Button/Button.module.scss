@import '../../_styles/theme';

$button-color: $color-primary !default;
$button-bg: #fff !default;
$button-width: 50px !default;
$button-padding: 6px 8px !default;
$button-border-color: $color-border !default;
$button-border-width: 1px !default;
$button-border-radius: 4px !default;
$button-hover-bg: $color-primary !default;
$button-hover-color: $color-white !default;

.button {
  box-sizing: border-box;
  white-space: nowrap;
  display: inline-flex;
  position: relative;
  margin: 0;
  line-height: 1.75;
  text-decoration: none;
  border: none;
  outline: none;
  cursor: default;

  &:not(:disabled) {
    &:hover {
      cursor: pointer;
    }
  }

  &.hasLoader:not(:disabled):hover {
    cursor: default;
  }

  .iconAfter {
    margin-left: 5px;
  }
  .iconBefore {
    margin-right: 5px;
  }

  &,
  svg {
    //transition: all 0.2s ease-in;
  }
}

.default {
  padding: $button-padding;
  min-width: $button-width;
  border-radius: $button-border-radius;
  color: $button-color;
  text-transform: uppercase;
  border: $button-border-width solid $button-border-color;
  background-color: $button-bg;

  svg {
    fill: $button-color;
  }

  &:not(:disabled) {
    &:hover,
    &:active {
      background: $button-hover-bg;
      color: $button-hover-color;
      cursor: pointer;

      svg {
        fill: $button-hover-color;
      }
    }
    &:active {
      box-shadow: inset 1px 1px 1px rgba(0, 0, 0, 0.3);
    }
  }
}

.cta {
  @extend .default;
  background-color: $color-primary;
  color: #fff;
}

.disabled {
  opacity: 0.4;
  color: $color-text-light;
  cursor: not-allowed;
}
