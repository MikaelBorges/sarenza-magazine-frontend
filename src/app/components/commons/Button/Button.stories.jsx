import '../../App.css';

import React from 'react';

import * as ICONS from '../Icon/constants';
import Button from './Button';

export default {
  title: 'Button',
  component: Button
};

export const Basic = () => <Button onClick={() => {}}>Button</Button>;
export const WithIcon = () => (
  <>
    <div className="storyBox">
      <Button onClick={() => {}} iconName={ICONS.CHEVRON_RIGHT}>
        Button
      </Button>
    </div>
    <div className="storyBox">
      <Button onClick={() => {}} iconName={ICONS.CHEVRON_RIGHT} iconAfter>
        Button
      </Button>
    </div>
    <div className="storyBox">
      <Button onClick={() => {}} iconName={ICONS.CHEVRON_DOWN} iconAfter>
        Button
      </Button>
    </div>
  </>
);
export const WithLoader = () => (
  <>
    <div className="storyBox">
      <Button onClick={() => {}} isLoading>
        Button
      </Button>
    </div>
    <div className="storyBox">
      <Button onClick={() => {}} isLoaded>
        Button
      </Button>
    </div>
  </>
);
