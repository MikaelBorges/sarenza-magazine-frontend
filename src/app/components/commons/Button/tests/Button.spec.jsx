import { render } from '@testing-library/react';
import React from 'react';

import Button from '../Button';

describe('Button', () => {
  it('should match the expected snapshot with default props', () => {
    const { container } = render(<Button />);

    expect(container).toMatchSnapshot();
  });

  it('should match the expected snapshot with complete props', () => {
    const props = {
      onClick: () => {},
      componentCustomClass: 'tunnel',
      title: 'toto'
    };
    const { container } = render(<Button {...props}>Plop</Button>);

    expect(container).toMatchSnapshot();
  });
});
