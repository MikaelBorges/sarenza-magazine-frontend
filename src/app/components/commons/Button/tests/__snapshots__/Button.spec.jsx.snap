// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match the expected snapshot with complete props 1`] = `
<div>
  <button
    class="button default status undefined"
    data-testid=""
    title="toto"
    type="button"
  >
    <span
      class="buttonLabel"
    >
      Plop
    </span>
  </button>
</div>
`;

exports[`Button should match the expected snapshot with default props 1`] = `
<div>
  Lien en erreur
</div>
`;
